src/client.rs

```diff
@@ -282,12 +325,27 @@ impl Response {
     }
 }
 
+/// Read the body of the Response, decoding it if necessary.
+impl Read for Decoder {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match self {
+            &mut Decoder::PlainText(ref mut hyper_response) => {
+                hyper_response.read(buf)
+            },
+            &mut Decoder::Gzip(ref mut hyper_response) => {
+                GzDecoder::new(BufReader::new(hyper_response))?.read(buf)
```

Comment:
Question: So if I am storing the ::hyper::client::Response object insider the decoder, I don't believe there is a way for me to access it because constructor the decoder takes ownership of it. But since `::reqwest::client::Response` still needs to support the `status()`, `headers()`, and `version()` methods we still need access to that data in some form. As I see it currently, we can change the definitions of the `Decoder::Gzip` variant to be something like:

```rust
enum Decoder {
    
    PlainText(::hyper::client::Response),
    
    Gzip {
        decoder: ::flate2::read::GzDecoder<::hyper::client::Response>,
        headers: ::hyper::headers::Headers,
        version: ::hyper::version::HttpVersion,
        status: ::hyper::status::StatusCode,
    }
}
```

and then we just do something like this where we copy the data before the request object ownership is lost.

```rust
    fn from_hyper_response(res: ::hyper::client::Response) -> Self {

        // ...

        if is_gzip {
            return Decoder::Gzip {
                status: res.status.clone(),
                version: res.version.clone(),
                headers: res.headers.clone(),
                decoder: GzDecoder::new(res).unwrap())
            };
        } else {
            return Decoder::PlainText(res);
        }
    }
```

but this strikes me as a poor solution. Storing fields named status, version, headers, on type named Decoder::Gzip seems incorrect to me. I actually think that data might be better off stored directly on the `::reqwest::client::Request` object, and then the inner `::hyper::client::Request` isn't stored inside the reqwest struct at all, instead maybe we just keep handle to the stream or a Read object of some sort.